autoware_lanelet2_map_validator/src/validators/intersection/intersection_area_tagging.cpp

```diff
@@ -0,0 +1,116 @@
+#include <lanelet2_core/primitives/Polygon.h>
+
+#include <map>
+#include <string>
+#include <unordered_set>
+
+namespace lanelet::autoware::validation
+{
+namespace
+{
+lanelet::validation::RegisterMapValidator<IntersectionAreaTaggingValidator> reg;
+}
+
+lanelet::validation::Issues IntersectionAreaTaggingValidator::operator()(
+  const lanelet::LaneletMap & map)
+{
+  lanelet::validation::Issues issues;
+
+  lanelet::autoware::validation::appendIssues(issues, check_intersection_area_tagging(map));
+
+  return issues;
+}
+
+lanelet::validation::Issues IntersectionAreaTaggingValidator::check_intersection_area_tagging(
+  const lanelet::LaneletMap & map)
+{
+  lanelet::validation::Issues issues;
+
+  std::map<lanelet::Id, lanelet::ConstPolygon3d> intersection_areas;
+  for (const lanelet::ConstPolygon3d & polygon3d : map.polygonLayer) {
+    if (
+      polygon3d.hasAttribute(lanelet::AttributeName::Type) &&
+      polygon3d.attribute(lanelet::AttributeName::Type).value() == "intersection_area") {
+      intersection_areas[polygon3d.id()] = polygon3d;
+    }
+  }
+
+  for (const auto & [area_id, area_polygon] : intersection_areas) {
+    lanelet::BasicPolygon2d area_polygon2d = lanelet::traits::to2D(area_polygon.basicPolygon());
+
+    for (const lanelet::ConstLanelet & lanelet : map.laneletLayer) {
+      lanelet::BasicPolygon2d lanelet_polygon = lanelet.polygon2d().basicPolygon();
+      if (boost::geometry::within(lanelet_polygon, area_polygon2d)) {
+        std::string intersection_area_attr = lanelet.attributeOr("intersection_area", "none");
+        if (intersection_area_attr == "none") {
+          // Issue-001: Lanelet missing intersection_area tag
+          std::map<std::string, std::string> area_id_map;
+          area_id_map["area_id"] = std::to_string(area_id);
+          issues.emplace_back(
+            construct_issue_from_code(issue_code(this->name(), 1), lanelet.id(), area_id_map));
+        } else {
+          lanelet::Id tagged_area_id =
+            static_cast<lanelet::Id>(std::atoi(intersection_area_attr.c_str()));
+          if (tagged_area_id != area_id) {
+            // Issue-002: Lanelet has wrong intersection_area tag
+            std::map<std::string, std::string> tag_map;
+            tag_map["expected_area_id"] = std::to_string(area_id);
+            tag_map["actual_area_id"] = intersection_area_attr;
+            issues.emplace_back(
+              construct_issue_from_code(issue_code(this->name(), 2), lanelet.id(), tag_map));
+          }
+        }
+      }
+    }
+  }
+
+  // Issue-003: Lanelet has intersection_area tag but is not completely covered by the referenced
+  // area
+  for (const lanelet::ConstLanelet & lanelet : map.laneletLayer) {
+    std::string intersection_area_attr = lanelet.attributeOr("intersection_area", "none");
+    if (intersection_area_attr != "none") {
+      lanelet::Id tagged_area_id =
+        static_cast<lanelet::Id>(std::atoi(intersection_area_attr.c_str()));
```

Comment:
I believe `attributeOr` also works with `lanelet::Id`
```suggestion
    lanelet::id tagged_area_id = lanelet.attributeOr("intersection_area", lanelet::InvalId);
    if (tagged_area_id != lanelet::InvalId)
```